File: day3/puzzle.py
import re 

import solver

get_mults_pattern = re.compile("(mul\([0-9]{1,3},[0-9]{1,3}\))")
get_digits_pattern = re.compile("[0-9]{1,3}")

## Puzzle 1
def read_str(file_name):
    with open(file_name, 'r') as file:
        str = file.read()
    
    return str

def do_multiplication(mul_str):
    nums = re.findall(get_digits_pattern, mul_str)
    return int(nums[0]) * int(nums[1])

def solve_puzzle1(str):
    matches = re.findall(get_mults_pattern, str)
    sum = 0 
    for match in matches:
        sum += do_multiplication(match)
        
    return sum

puzzle1_solver = solver.Solver("regex", 3, 1, read_str, solve_puzzle1)

## Puzzle 2
get_mults_and_conds_pattern = re.compile("(mul\([0-9]{1,3},[0-9]{1,3}\))|(do\(\))|(don\'t\(\))")

def solve_puzzle2(str):
    matches = re.findall(get_mults_and_conds_pattern, str)
    sum = 0
    is_enabled = True
    for match in matches:
        if match[1]:
            is_enabled = True 
        elif match[2]:
            is_enabled = False
        elif is_enabled:
            sum += do_multiplication(match[0])

    return sum

puzzle2_solver = solver.Solver("regex", 3, 2, read_str, solve_puzzle2)

File: day3/puzzle.cpp
#include <sstream>
#include <regex>

#include "../headers/solver.h"

int solve_puzzle_1(std::string str)
{
    std::regex mults_regex("(mul\\([0-9]{1,3},[0-9]{1,3}\\))");
    

// get_mults_pattern = re.compile("")
// get_digits_pattern = re.compile("[0-9]{1,3}")
}

namespace day3
{
    class Day3Solver : Solver<std::string, int>
    {
    public:
        Day3Solver(std::string name, int puzzle_index) : Solver(name, 3, puzzle_index) {}

        std::string parseInputFile(std::string file_name) override
        {
            std::ifstream file;
            file.open(file_name);

            std::stringstream sstream;
            sstream << file.rdbuf();
            std::string str = sstream.str();
            return str;
        }
    };

    class Puzzle1Solver : Day3Solver
    {
    public:
        Puzzle1Solver() : Day3Solver("regex", 1) {}

        int computeSolution(std::string data) override
        {
            return 0;
        }
    };
}

